### PR TITLE
shutdown strategy

### PR DESCRIPTION
**Pull request by tdantas**
# Description

- Loop/Wait for all workers on shutdown
- Leave any jobs as-is (eg: pending) on shutdown

### CODE REVIEW

File: src/ablauf/job/sql.clj [hunk 1]
```diff
@@ -219,15 +232,44 @@
...
+          ;; now we need to mark the proper job status
+          (jdbc/with-transaction [tx db]
+            ;; the process-available loop will always insert the next task transactionally
+            ;; if there is one
+            (when-let [next-task (next-task? tx (:task/wid processed-task))]
+              (let [{:task/keys [wid payload]} next-task
+                    workflow (workflow-by-id tx wid)
+                    [ast context _actions] (job/restart (:workflow_run/job workflow) payload)
+                    aborted-ast           (job/abort ast)]
+                (update-workflow tx wid aborted-ast context))))
```
Comment (mping-exo) [inline src/ablauf/job/sql.clj:255]: By marking the job as aborted, we will require further processing if we want to be able to replay the job.

Another alternative for shutdown is to just let the workflow run marked as pending, and on process restart, we can scan pending jobs and decide what to do. If a pending job is lockable, it means there is no other process hanging on to the job and we can pick it up again.

File: src/ablauf/job/sql.clj [hunk 2]
```diff
@@ -219,15 +232,44 @@
...
+          (do
+            (Thread/sleep (+ wait-ms (rand-int wait-allowance)))
+            (recur)))))))
+
+(defn shutdown!
+  "Sets the system in shutdown state:
+  - pending leafs will be run to completion
+  - new jobs will be rejected"
+  []
+  (swap! state assoc :shutdown? true))
```
Comment (mping-exo) [inline src/ablauf/job/sql.clj:267]: Should we wait for pending jobs to finish before returning?

File: src/ablauf/job/sql.clj (hunk 1 above)
Comment (mping-exo) [reply src/ablauf/job/sql.clj]: I ended up reverting and using the simplest option of not doing nothing

File: src/ablauf/job/sql.clj (hunk 2 above)
Comment (mping-exo) [reply src/ablauf/job/sql.clj]: Added loop for waiting for all workers before returning

**Review by mping-exo — APPROVED** [review]

**Review by wdullaer — APPROVED** [review]

File: src/ablauf/job/sql.clj [hunk 3]
```diff
@@ -65,6 +68,14 @@
             (update :task/wuuid parse-uuid)
             (assoc :task/status "pending"))))
 
+(defn- unpack-workflow-run [{:workflow_run/keys [reason ast context] :as wrun}]
+  (-> wrun
```
[the remark this answers is not in the document]
Comment (mping-exo) [reply src/ablauf/job/sql.clj:72]: Yes, its a DB row. If it's null its better to throw and fix the bug than silently proceed

**Review by bcachet — APPROVED** [review]
May I ask why the _worker_ mecanism in `ablauf.job.sql` is not composed of a thread-pool (fixed or not) on top of which we submit some operation to be performed.
On shutdown we can shutdown the thread pool and await for termination ?

**Comment by mping-exo** [comment]
@bcachet I think it was by design (initial design was by @pyr). 
The threadpool can be later configured in the calling system (eg: https://github.com/exoscale/root-api/blob/master/modules/root-api-server/src/root_api/runner.clj#L223), and ablauf just exposes a way to create a worker.

**Comment by mping-exo** [comment]
> > @bcachet I think it was by design (initial design was by @pyr).
> > The threadpool can be later configured in the calling system (eg: https://github.com/exoscale/root-api/blob/master/modules/root-api-server/src/root_api/runner.clj#L223), and ablauf just exposes a way to create a worker.
> 
> In this case why not handling shutdown at the thread pool level in root-api ? Root-api should be responsible of not accepting new operation by not listening to new request on its socket instead of refactoring to throw exception when receiving request while shutting down

In retrospect, I think you are right. I think the PR was created before the roo-api integration, but now it doesn't make sense anymore, since we assume the consumer of the library will be handling its own threadpool.
Closing

**Pull request closed without merging**